core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -702,31 +644,171 @@ public String toString()
+        {
+            Domain result;
+            synchronized (this) {
+                if (collected) {
+                    return;
+                }
+                collectedTaskCount++;
+                collected = true;
+                result = domain;
+            }
+            collectionDuration.set(Duration.succinctNanos(System.nanoTime() - start));
+            collectedDomainsFuture.set(result);
+        }
+
+        private void collectPartitioned(TaskId taskId, Domain domain)
+        {
+            if (!collectedTasks.add(taskId.getPartitionId())) {
+                return;
+            }
+            summaryDomains.add(domain);
+            unionSummaryDomains();
+
+            Domain result;
+            synchronized (this) {
+                if (collected) {
+                    return;
+                }
+                collectedTaskCount++;
+                boolean allPartitionsCollected = expectedTaskCount != null && expectedTaskCount == collectedTaskCount;
+                if (allPartitionsCollected) {
+                    // run final compaction as previous concurrent compactions may have left more than a single domain
+                    unionSummaryDomains();
+                }
+
+                boolean collectionFinished = domain.isAll() || allPartitionsCollected;
+                if (!collectionFinished) {
+                    return;
+                }
+                collected = true;
+                if (domain.isAll()) {
+                    result = domain;
+                }
+                else {
+                    // run union one more time
+                    unionSummaryDomains();
+                    int summaryDomainsCount = summaryDomains.size();
+                    verify(summaryDomainsCount == 1, "summaryDomainsCount is expected to be equal to 1, got: %s", summaryDomainsCount);
+                    result = summaryDomains.poll();
+                }
+            }
+
+            verify(result != null);
+            collectionDuration.set(Duration.succinctNanos(System.nanoTime() - start));
+            collectedDomainsFuture.set(result);
+        }
+
+        private void unionSummaryDomains()
+        {
+            while (true) {
+                // This method is called every time a new domain is added to the summaryDomains queue.
+                // In a normal situation (when there's no race) there should be no more than 2 domains in the queue.
+                Domain first = summaryDomains.poll();
+                if (first == null) {
+                    return;
+                }
+                Domain second = summaryDomains.poll();
+                if (second == null) {
+                    summaryDomains.add(first);
+                    return;
+                }
+                summaryDomains.add(first.union(second));
+            }
+        }
+
+        public void setExpectedTaskCount(int count)
+        {
+            if (collected || expectedTaskCount != null) {
```

Comment:
Should this condition be a `checkState` or is this a valid scenario ?

Reply:
With the current implementation it is possible for this method to be called more than once. 

To be honest I'm not a huge fan of the current flow (`stageDynamicFilters` + `stageNumberOfTasks`) and likely with the future iterations of fault tolerant execution when fragments and stages no longer correspond to 1-1 relation it might not even work. So probably in the future we might want to make this explicit (when scheduler set's expected dynamic filters and the expected number of tasks directly). But I decided to postpone this refactor until necessary.